Assignments/A3/Distro/QLearn.c

```diff
@@ -416,3 +440,88 @@ int pos_to_index(int pos[2], int size_X) { return pos[0] + pos[1] * size_X; }
 int is_pos_valid(int pos[2], int size_X, int size_Y) {
   return 0 <= pos[0] && pos[0] < size_X && 0 <= pos[1] && pos[1] < size_Y;
 }
+
+// Avg Cat Dist Feature
+double avg_cat_feat(double gr[max_graph_size][4], double features[25],
+                    int mouse_pos[1][2], int cats[5][2], int size_X,
+                    int graph_size) {
+  // -1 is terrible, aka cats are right on top
+  // 1 is great, aka cats are across the map from the mouse
+  double max_dist = pow(
+      pow((double)(size_X), 2) + pow((double)(graph_size / size_X), 2), 0.5);
+  double average_cat_distance = 0;
+  int num_cat = 0;
+  while (cats[num_cat][0] != -1) {
+    const double kCatDistance =
+        pow(pow((double)(mouse_pos[0][0] - cats[num_cat][0]), 2) +
+                pow((double)(mouse_pos[0][1] - cats[num_cat][1]), 2),
+            0.5);
+    average_cat_distance += kCatDistance;
+    ++num_cat;
+  }
+  // Get avg distance and normalize on max distance to [0,1]
+  average_cat_distance /= num_cat * max_dist;
+  // manipulate range to [-0.5, 0.5] *2 = [-1, 1]
+  return (average_cat_distance - 0.5) * 2;
+}
+
+// Closest Cat Dist Feature and Closest Cheese Dist Feature
+double closest_dist(double gr[max_graph_size][4], double features[25],
+                    int mouse_pos[1][2], int agents[5][2], int size_X,
+                    int graph_size) {
+  // -1 is terrible, aka cats are right on top
+  // 1 is great, aka cats are across the map from the mouse
+  double max_dist = pow(
+      pow((double)(size_X), 2) + pow((double)(graph_size / size_X), 2), 0.5);
+  double closest_dist = max_dist;
+  int num_agent = 0;
+  while (agents[num_agent][0] != -1) {
+    double dist =
+        pow(pow((double)(mouse_pos[0][0] - agents[num_agent][0]), 2) +
+                pow((double)(mouse_pos[0][1] - agents[num_agent][1]), 2),
+            0.5);
+    closest_dist = fmin(dist, closest_dist);
+    ++num_agent;
+  }
+  closest_dist /= max_dist;
+  return (closest_dist - 0.5) * 2;
+}
+
+// Check if currenty at dead end
+bool deadEnd(double gr[max_graph_size][4], int mouse_pos[1][2], int size_X) {
+  int walls = 0;
+  for (int i = 0; i < 4; ++i) {
+    walls += gr[pos_to_index(mouse_pos[0], size_X)][i];
+  }
+  if (walls > 2) {
+    return (true);
+  }
+  return (false);
+}
+
+// Helper function for angle
+double length(Point p1, Point p2) {
+  return sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y));
+}
+
+// Define a function that calculates the angle between three points
+double angle(int mouse_pos[1][2], int cats[5][2], int cheese[2]) {
+  int num_c = 0, c_x = 0, c_y = 0;
+  while (cats[num_c][0] != -1) {
+    c_x += cats[num_c][0];
+    c_y += cats[num_c][1];
+    ++num_c;
+  }
```

Comment:
Could be a for-loop instead:
```
for (int i=0; cats[i][0] != -1; ++i) {
  ...
}
```

Reply:
I need this for `line 516`, since I divide by `num_c` to get the avg cat position

Reply:
In this case you could do:
```c
int num_cats = 0;
for (int num_cats=0; cats[num_cats][0] != -1; ++num_cats) {
  ...
}
// num_cats will be what you want here.
```

Reply:
Can you explain why you don't want a while loop? Just curious why a for loop would be better even though the lines of code would be basically the same

Reply:
Yea fixed the two declaration thing.
Using `for (int i=0; [...]; ++i)` is a standard format - meaning a C programmer could understand what it means at a glance.
On the other hand, using a `while` loop spreads this logic over multiple lines.

Reply:
Cool, thanks :)